plasmapy/simulation/particle_tracker.py

```diff
@@ -0,0 +1,440 @@
+                _Ex, _Ey, _Ez, _Bx, _By, _Bz = grid.volume_averaged_interpolator(
+                    pos,
+                    "E_x",
+                    "E_y",
+                    "E_z",
+                    "B_x",
+                    "B_y",
+                    "B_z",
+                    persistent=True,
+                )
+            elif self.field_weighting == "nearest neighbor":
+                _Ex, _Ey, _Ez, _Bx, _By, _Bz = grid.nearest_neighbor_interpolator(
+                    pos,
+                    "E_x",
+                    "E_y",
+                    "E_z",
+                    "B_x",
+                    "B_y",
+                    "B_z",
+                    persistent=True,
+                )
+
+            # Interpret any NaN values (points off the grid) as zero
+            # Do this before adding to the totals, because 0 + nan = nan
+            _Ex = np.nan_to_num(_Ex, nan=0.0 * u.V / u.m)
+            _Ey = np.nan_to_num(_Ey, nan=0.0 * u.V / u.m)
+            _Ez = np.nan_to_num(_Ez, nan=0.0 * u.V / u.m)
+            _Bx = np.nan_to_num(_Bx, nan=0.0 * u.T)
+            _By = np.nan_to_num(_By, nan=0.0 * u.T)
+            _Bz = np.nan_to_num(_Bz, nan=0.0 * u.T)
+
+            # Add the values interpolated for this grid to the totals
+            Ex += _Ex
+            Ey += _Ey
+            Ez += _Ez
+            Bx += _Bx
+            By += _By
+            Bz += _Bz
+
+        # Create arrays of E and B as required by push algorithm
+        E = np.array(
+            [Ex.to(u.V / u.m).value, Ey.to(u.V / u.m).value, Ez.to(u.V / u.m).value]
+        )
+        E = np.moveaxis(E, 0, -1)
+        B = np.array([Bx.to(u.T).value, By.to(u.T).value, Bz.to(u.T).value])
+        B = np.moveaxis(B, 0, -1)
+
+        # Calculate the adaptive timestep from the fields currently experienced
+        # by the particles
+        # If user sets dt explicitly, that's handled in _adaptive_dt
+        dt = self._adaptive_dt(Ex, Ey, Ez, Bx, By, Bz)
+
+        # TODO: Test v/c and implement relativistic Boris push when required
+        # vc = np.max(v)/_c
+
+        # If dt is not a scalar, make sure it can be multiplied by an
+        # [nparticles, 3] shape field array
+        if dt.size > 1:
+            dt = dt[:, np.newaxis]
+
+        self.x, self.v = boris_push(
+            self.x, self.v, B, E, self.q, self.m, dt, inplace=False
+        )
+
+    @property
+    def on_any_grid(self):
+        """
+        Binary array for each particle indicating whether it is currently
+        on ANY grid.
+        """
+        return np.sum(self.on_grid, axis=-1) > 0
+
+    @staticmethod
+    def _stop_condition():
+        return False
+
+    @property
+    def vmax(self):
+        """
+        Calculate the maximum velocity.
+        Used for determining the grid crossing maximum timestep.
+        """
+        return np.max(np.linalg.norm(self.v, axis=-1))
+
+    def _validate_inputs(self, field_weighting: str):
```

Comment:
I'm not sure this really belongs in it's own function: it's only called once. If so, I would call it `_validate_inputs_run` and have it handle all the arguments for `run`.  Or `_validate_field_weighting_str` if it only does this

Reply:
It's called first inside the function body for `run` defined in both classes. I was running into an issue with the test cases where the radiography `run` would stop some particles causing the `grid_ind` mask to not represent all of the cases not being tested. I converted the check into a function to ensure we are testing the value of field weighting at the beginning of both `run` functions for consistency. I agree, `_validate_field_weighting_str` would probably be a better name though